pkg/apis/garden/v1beta1/types.go

```
@@ -723,9 +723,9 @@ type ShootStatus struct {
 	// LastOperation holds information about the last operation on the Shoot.
 	// +optional
 	LastOperation *gardencorev1alpha1.LastOperation `json:"lastOperation,omitempty"`
-	// LastError holds information about the last occurred error during an operation.
+	// LastErrors holds information about the last occurred error(s) during an operation.
 	// +optional
-	LastError *gardencorev1alpha1.LastError `json:"lastError,omitempty"`
+	LastErrors []gardencorev1alpha1.LastError `json:"lastErrors,omitempty"`
```

This change might cause issues....
I mean the API might not start up because it cannot decode existing objects.

---

Testing it locally didn't cause errors. I guess because both are optional the status update simply removes one and adds the other.

---

I am reading https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api_changes.md and still wondering whether this change is acceptable.

There is even example with switching to array.
```golang
// Still API v6.
type Frobber struct {
  Height int           `json:"height"`
  Width  int           `json:"width"`
  Param  string        `json:"param"`  // the first param
  Params []string      `json:"params"` // all of the params
}
```


I think that both fields should be presented in the v1beta1 API with the old one marked as deprecated and removed after couple of releases.
Do you think a conversion between both fields is feasible?
Basically something like this:
```golang
if LastError != nil {
    LastErrors = append(LastErrors, *LastError)
}
```

@mvladev @rfranzke wdyt?

---

I don't think it's acceptable. I don't even understand why "testing it locally didn't cause errors". Can you explain that? Let's assume a shoot has a last error, the json will look like this:

`{..., "lastError: {"description": "foobar"}}`

Now, when trying to unmarshal this into `[]LastError` it will fail, right?

---

The field name is changed, too `lastError` -> `lastErrors`, so they are completely different fields. For sure data will be lost. But I am not sure whether it should fail if this field is presented in the etcd, but is not defined in the API?

---

It's simply ignored if it's in etcd but not in the API. However, wrt backwards compatibility I guess we can't do this either.

---

You're right that both should be present. The data loss can have an effect on the shoot care controller which I didn't take into account: if there already exist a LastError and it is not added to the LastErrors array the following code will act as if there are no errors: [shoot_care_control.go:177](https://github.com/gardener/gardener/blob/1ed7cdbbbbeb739b91fffb4f5c7d586c61a2815d/pkg/controllermanager/controller/shoot/shoot_care_control.go#L177). 
Additionally the transition to the new array field for the gardener dashboard will not have to happen immediately after it is added.

Thanks for the review and @vpnachev for pointing me to that documentation.

---

@vpnachev @rfranzke After thinking a bit about the backwards compatibility there could be the following issues:

If we keep only `LastErrors`(plural) in the internal representation of `ShootStatus` and both `LastError` and `LastErrors` in the versioned APIs: 
1. During conversion from internal to versioned to create `LastError`(singular) there is no way to construct the `LastError.Description` the same as it would have been before this change (unless we get it from the `LastOperation.Description`).
2. During updates which first retrieve the `ShootStatus` and then only change either `LastError` or `LastErrors`(plural) we have to use `LastError.LastUpdateTime` to decide which one to add to the internal representation and then save.

---

What I could do is to just change the LastError API by adding a new optional `TaskErrors` field:
```
type LastError {
    Description string
    // optional
    TaskID string
    // optional
    TaskErrors []TaskError
    // optional
    Codes []ErrorCodes
}

type TaskError {
   Description string
   // optional
   TaskID string
   // optional
   Codes []ErrorCodes
}
```

Two examples would be:
```
// When the error is generated during a flow execution
LastError {
    Description: Flow reconcile encountered task errors
    TaskErrors: [
        TaskError {
            Description: Faied to deploy network extension
            TaskID: Deploy network extension
        }

        TaskError {
            Description: Failed to connect to shoot apiserver
            TaskID: Connect to shoot apiserver
            Codes: "404"
        }
    ]
}

// when the error was not generated by the flow
LastError {
    Description: Failed to check versions
    TaskID: CheckVersions
}
```

When a step with a `taskID` succeeds and we have the ErrorCleaner configured:
1. Find the LastError or TaskError which has the same `taskID`.
2. If `taskID` matches the TaskID of the LastError remove it completely.
2. For the TaskErrors Create a remainingErrors array and only add TaskErrors which do not match the given `taskID` to the array
3. If len(remainingErrors) == 0 and len(LastErrors.TaskErrors) > 0 clean up LastError as well. This comparison is to avoid deleting errors which have neither TaskID nor TaskErrors.

@rfranzke @mvladev @vpnachev wdyt?